`@CompileStatic` class fails to compile with Groovy 3.0.21 shipped in 8.8

### Current Behavior

This simple class doesn't compile with Gradle 8.8, which worked in previous versions.

```
import org.gradle.api.tasks.Exec


@CompileStatic
class PromotionPlugin implements Plugin<Project> {
    private Project project;

    @Override
    void apply(Project project) {
        this.project = project;
    }

    private Exec createGitTask(String name, List args) {
        return project.tasks.create(name, Exec, { Exec it ->
            it.executable "git"
            it.args(args)
        } as Action)
    }
}
```

### Expected Behavior

It compiles.

### Context (optional)

See the section below for a reproducer. 

Bisect to https://github.com/gradle/gradle/compare/1d270aa1c218e63564b7aba1519e3d1437252075...2690530cc2812d07c6f8a565023cb79d4c688c62

It's very likely to be Groovy 3.0.21 introduces this breakage. There seems no workaround if the class is in `buildSrc` - we can't even downgrade to 3.0.18 because of versioin conflict:

```
Caused by: groovy.lang.GroovyRuntimeException: Conflicting module versions. Module [groovy-datetime is loaded in version 3.0.21 and you are trying to load version 3.0.18
```

### Steps to Reproduce

Run `./gradlew help` in [gradle-8.8-compile-static-reproducer.zip](https://github.com/user-attachments/files/15528387/gradle-8.8-compile-static-reproducer.zip)

```
$ ./gradlew help

> Task :buildSrc:compileGroovy FAILED
startup failed:
/Users/bo/Projects/tmp/gradle-8.8-compile-static-reproducer/buildSrc/src/main/groovy/org/my/PromotionPlugin.groovy: 22: [Static type checking] - Cannot find matching method java.lang.Object#executable(java.lang.String). Please check if the declared type is correct and if the method exists.
 @ line 22, column 13.
               it.executable "git"
               ^

/Users/bo/Projects/tmp/gradle-8.8-compile-static-reproducer/buildSrc/src/main/groovy/org/my/PromotionPlugin.groovy: 23: [Static type checking] - Cannot find matching method java.lang.Object#args(java.util.List). Please check if the declared type is correct and if the method exists.
 @ line 23, column 13.
               it.args(args)
               ^

2 errors


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildSrc:compileGroovy'.
> Compilation failed; see the compiler error output for details.
```



### Gradle version

8.8

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
+++ b/platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc
@@ -327,6 +327,15 @@ Groovy has been updated to https://groovy-lang.org/changelogs/changelog-3.0.21.h
 
 Since the previous version was 3.0.17, the https://groovy-lang.org/changelogs/changelog-3.0.18.html[3.0.18] and https://groovy-lang.org/changelogs/changelog-3.0.19.html[3.0.19], and https://groovy-lang.org/changelogs/changelog-3.0.20.html[3.0.20] changes are also included.
 
+Some changes in static type checking have resulted in source-code incompatibilities. One of these occurs if you were casting a closure to an `Action` without generics. Starting with 3.0.18, this now results in the parameter of the closure being `Object` instead of any explicit type specified. This can be fixed by adding the appropriate type to the cast, and the redundant parameter declaration can be removed:
+```groovy
+// Before
+tasks.create("foo", { Task it -> it.description "Foo task" } as Action)
+
+// Fixed
+tasks.create("foo", { it.description "Foo task" } as Action<Task>)
+```
+
 ==== Upgrade to ASM 9.7
 
 ASM was upgraded from 9.6 to https://asm.ow2.io/versions.html[9.7] to ensure earlier compatibility for Java 23.

